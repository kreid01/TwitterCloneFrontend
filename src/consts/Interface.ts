export type IPost = {
  id: number;
  userName: string | undefined;
  userAt: string | undefined;
  userImg: string | undefined;
  postTextBody: string;
  postMedia: string;
  postDate: string;
  commentCount: number;
  likeCount: number;
  retweetCount: number;
  isLiked?: boolean;
  isRetweeted?: boolean;
  comments?: Array<Comment>;
  posterId?: number;
  likedBy?: number[];
  retweetedBy?: number[];
};

export type IComment = {
  id: number;
  userAt: string | undefined;
  userName: string | undefined;
  userImg: string | undefined;
  commentBody: string;
  commentMedia?: string;
  commentDate: string;
};

export type INewComment = {
  commentBody: string;
  commentMedia: string;
  userAt: string | undefined;
  userName: string | undefined;
  userImg: string | undefined;
};

export type INewPost = {
  postTextBody: string;
  postMedia: string;
  userAt: string | undefined;
  userName: string | undefined;
  userImg: string | undefined;
};

export type IUser = {
  userId: number;
  userName: string;
  userAt: string;
  userImg: string;
  userCoverImg: string;
  userEmail: string;
  joinDate: string;
  followers: Array<number>;
  following: Array<number>;
  isAdmin?: boolean;
  userPassword?: string;
} | null;
