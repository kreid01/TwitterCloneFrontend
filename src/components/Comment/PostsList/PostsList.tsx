import React, { useState } from "react";
import { CurrentPost } from "components/Comment/CurrentPost/CurrentPost";
import { Post } from "components/Post/Post";
import { nanoid } from "nanoid";

import { IPost } from "../../../consts/Interface";
import {
  useIsCommenting,
  useUpdateIsCommenting,
} from "context/IsCommentingContext";

interface Props {
  setPosts: React.Dispatch<React.SetStateAction<IPost[] | undefined>>;
  hasMore: boolean;
  posts: IPost[];
  loading: boolean;
  error: boolean;
  loader: React.MutableRefObject<null>;
  userId?: number;
}

export const PostsList: React.FC<Props> = ({
  setPosts,
  hasMore,
  posts,
  loading,
  error,
  loader,
  userId,
}) => {
  const postsList = () => {
    if (typeof posts !== "undefined") {
      return posts.map((post, index) => {
        let isUsersPost = false;
        if (post.posterId === userId) isUsersPost = true;
        return (
          <div>
            <Post
              isUsersPost={isUsersPost}
              post={post}
              setPosts={setPosts}
              posts={posts}
              key={nanoid()}
              makeCurrentPost={makeCurrentPost}
            />
          </div>
        );
      });
    }
  };
  const [currentPost, setCurrentPost] = useState<IPost>();
  const toggleIsCommenting = useUpdateIsCommenting();
  const isCommenting = useIsCommenting();

  const makeCurrentPost = (post: IPost) => {
    setCurrentPost(post);
    if (toggleIsCommenting !== null) toggleIsCommenting();
  };

  return isCommenting && currentPost !== undefined ? (
    <>
      <CurrentPost
        makeCurrentPost={makeCurrentPost}
        isUsersPost={false}
        posts={posts}
        setPosts={setPosts}
        post={currentPost as IPost}
      />
    </>
  ) : (
    <>
      <div className="container-fluid">
        <div>{postsList()}</div>
      </div>
      {loading && hasMore && <p className="pl-4 pt-4">Loading...</p>}
      {error && <p>Error!</p>}
      <div ref={loader} />
    </>
  );
};
