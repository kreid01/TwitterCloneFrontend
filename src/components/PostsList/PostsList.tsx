import React, { useEffect, useState } from "react";
import { CurrentPost } from "components/Comment/CurrentPost/CurrentPost";
import { Post } from "components/Post/Post";
import { nanoid } from "nanoid";

import { IPost } from "../../consts/Interface";
import {
  useIsCommenting,
  useUpdateIsCommenting,
} from "context/IsCommentingContext";
import { useGetUser } from "context/UserContext";

interface Props {
  setPosts: React.Dispatch<React.SetStateAction<IPost[] | undefined>>;
  hasMore: boolean;
  posts: IPost[];
  loading: boolean;
  error: boolean;
  loader?: React.MutableRefObject<null>;
}

export const PostsList: React.FC<Props> = ({
  setPosts,
  hasMore,
  posts,
  loading,
  error,
  loader,
}) => {
  const user = useGetUser();
  const [currentPost, setCurrentPost] = useState<IPost>();
  const toggleIsCommenting = useUpdateIsCommenting();
  const isCommenting = useIsCommenting();
  const [wait, setWait] = useState(false);

  setInterval(() => setWait(true), 500);

  useEffect(() => {
    if (typeof posts !== "undefined") {
      posts.map((post, index) => {
        if (post.likedBy?.includes(user?.userId as number)) {
          const newArr = [...posts];
          newArr[index].isLiked = true;
          setPosts(newArr);
        }
      });
    }
  }, [wait]);

  useEffect(() => {
    if (typeof posts !== "undefined") {
      posts.map((post, index) => {
        if (post.retweetedBy?.includes(user?.userId as number)) {
          const newArr = [...posts];
          newArr[index].isRetweeted = true;
          setPosts(newArr);
        }
      });
    }
  }, [wait]);

  const postsList = () => {
    if (typeof posts !== "undefined") {
      return posts.map((post, index) => {
        let isUsersPost = false;
        if (post.posterId === user?.userId) isUsersPost = true;
        return (
          <div>
            <Post
              isUsersPost={isUsersPost}
              post={post}
              setPosts={setPosts}
              posts={posts}
              key={nanoid()}
              makeCurrentPost={makeCurrentPost}
            />
          </div>
        );
      });
    }
  };

  const makeCurrentPost = (post: IPost) => {
    setCurrentPost(post);
    if (toggleIsCommenting !== null) toggleIsCommenting();
  };

  return isCommenting && currentPost !== undefined ? (
    <>
      <CurrentPost
        makeCurrentPost={makeCurrentPost as () => void}
        isUsersPost={false}
        posts={posts}
        setPosts={setPosts}
        post={currentPost as IPost}
      />
    </>
  ) : (
    <>
      <div className="container-fluid">
        <div key={nanoid()}>{postsList()}</div>
      </div>
      {loading && hasMore && <p className="pl-4 pt-4">Loading...</p>}
      {error && <p>Error!</p>}
      <div ref={loader} />
    </>
  );
};
