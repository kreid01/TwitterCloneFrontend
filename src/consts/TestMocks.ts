export const testPost = {
  id: 1,
  userName: "test-name",
  userAt: "testAt",
  userImg: "testImg",
  postTextBody: "testBody",
  postMedia: "test-media",
  postDate: "testDate",
  commentCount: 1,
  likeCount: 1,
  retweetCount: 1,
  isRetweeted: false,
  isLiked: false,
  retweetedBy: [1],
  comments: [],
};

export const testComment = {
  id: 1,
  userName: "test-name",
  userAt: "testAt",
  userImg: "testImg",
  commentBody: "testBody",
  commentMedia: "test-media",
  commentDate: "test-date",
};
export const testPosts = [testPost, testPost];

export const testUser = {
  userId: 1,
  userName: "test-name",
  userAt: "testAt",
  userImg: "testImg",
  userCoverImg: "testCoverImg",
  userEmail: "testEmail",
  postsIds: [1, 2],
  joinDate: "testDate",
  following: [1, 2],
  followers: [1, 2],
  email: "",
  password: "",
};
