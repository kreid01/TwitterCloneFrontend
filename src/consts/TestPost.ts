export const testPost = {
    id:1,
    userName:'testName',
    userAt:'testAt',
    userImg:'testImg',
    postTextBody:'testBody',
    postMedia:'testMedia',
    postDate:'testDate',
    commentCount:1,
    likeCount:1,
    retweetCount:1
}