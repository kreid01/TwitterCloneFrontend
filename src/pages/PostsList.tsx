import React from "react";
import { CurrentPost } from "components/Comment/CurrentPost/CurrentPost";
import { postPost } from "services/postPost";
import { CreatePost } from "components/NewPost/CreatePost/CreatePost";
import { Post } from "components/Post/Post";
import { nanoid } from "nanoid";

import { INewPost, IPost } from "../consts/Interface";

interface Props {
  handleLike: (
    posts: IPost[],
    index: number,
    setter: (setterArr: IPost[]) => void
  ) => void;
  handleRetweet: (
    posts: IPost[],
    index: number,
    setter: (setterArr: IPost[]) => void
  ) => void;
  handleComment: (post: IPost) => void;
  setToCurrentPost: (post: IPost, index: number) => void;
  closeComment: () => void;
  setPosts: React.Dispatch<React.SetStateAction<IPost[] | undefined>>;
  isCommenting: boolean;
  isOnCurrentPost: boolean;
  currentPost: IPost;
  currentIndex: number;
  hasMore: boolean;
  posts: IPost[];
  loading: boolean;
  error: boolean;
  loader: React.MutableRefObject<null>;
}

export const PostsList: React.FC<Props> = ({
  handleComment,
  handleLike,
  handleRetweet,
  setToCurrentPost,
  closeComment,
  setPosts,
  isCommenting,
  isOnCurrentPost,
  currentIndex,
  currentPost,
  hasMore,
  posts,
  loading,
  error,
  loader,
}) => {
  const postsList = () => {
    if (typeof posts !== "undefined") {
      return posts.map((post, index) => {
        return (
          <div>
            <Post
              handleLike={handleLike}
              handleRetweet={handleRetweet}
              handleComment={handleComment}
              setToCurrentPost={setToCurrentPost}
              post={post}
              setPosts={setPosts}
              posts={posts}
              key={nanoid()}
              index={index}
            />
          </div>
        );
      });
    }
  };

  return isOnCurrentPost ? (
    <>
      <CurrentPost
        currentIndex={currentIndex as number}
        setToCurrentPost={setToCurrentPost}
        posts={posts}
        setPosts={setPosts}
        handleLike={handleLike}
        handleRetweet={handleRetweet}
        handleComment={handleComment}
        closeComment={closeComment}
        isCommenting={isCommenting}
        post={currentPost as IPost}
      />
    </>
  ) : (
    <>
      <div className="container-fluid pl-5">
        <div>{postsList()}</div>
      </div>
      {loading && hasMore && <p>Loading...</p>}
      {error && <p>Error!</p>}
      <div ref={loader} />
    </>
  );
};
