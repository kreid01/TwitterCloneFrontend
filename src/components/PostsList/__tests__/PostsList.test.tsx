import React from "react";
import { PostsList } from "../PostsList";
import { testPosts } from "../../../consts/TestMocks";

import { screen, render, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

const mockFn = jest.fn;

afterEach(cleanup);
it("render without crashing", async () => {
  render(
    <BrowserRouter>
      <PostsList
        posts={testPosts}
        setPosts={mockFn}
        hasMore={true}
        loading={true}
        error={true}
        key={1}
      />
    </BrowserRouter>
  );
});

it("loading renders when needed", async () => {
  render(
    <BrowserRouter>
      <PostsList
        posts={testPosts}
        setPosts={mockFn}
        hasMore={true}
        loading={true}
        error={true}
        key={1}
      />
    </BrowserRouter>
  );
  const element = screen.getByText("Loading...");
  expect(element).toBeVisible();
});

it("matched snapshot", async () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <PostsList
          posts={testPosts}
          setPosts={mockFn}
          hasMore={true}
          loading={true}
          error={true}
          key={1}
        />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
