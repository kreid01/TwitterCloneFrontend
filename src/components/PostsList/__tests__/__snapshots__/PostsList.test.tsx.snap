// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matched snapshot 1`] = `
Array [
  <div
    className="container-fluid"
  >
    <div>
      <div>
        <div
          className=" hover:bg-neutral-100 border-b-2 px-4 pt-4"
        >
          <div
            className="flex"
            data-testid="post"
            onClick={[Function]}
          >
            <img
              alt=""
              className="w-12 h-12 rounded-full"
              src="testImg"
            />
            <div
              className="flex-col pl-4"
            >
              <p
                className=""
              >
                <a
                  href="/testAt"
                  onClick={[Function]}
                >
                  <strong>
                    test-name
                  </strong>
                </a>
                <span
                  className="text-gray-500"
                >
                  @
                  testAt
                  ·
                  testDat
                </span>
              </p>
              <p>
                testBody
              </p>
              <img
                alt=""
                className="w-80 pt-2 max-h-72 rounded-xl"
                data-testid="media"
                src="test-media"
              />
            </div>
          </div>
          <div
            className="ml-10"
          >
            <div
              className="flex"
            >
              <div
                className="comment-icon"
              >
                <svg
                  data-testid="commentButton"
                  fill="currentColor"
                  height="1em"
                  onClick={[Function]}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div>
        <div
          className=" hover:bg-neutral-100 border-b-2 px-4 pt-4"
        >
          <div
            className="flex"
            data-testid="post"
            onClick={[Function]}
          >
            <img
              alt=""
              className="w-12 h-12 rounded-full"
              src="testImg"
            />
            <div
              className="flex-col pl-4"
            >
              <p
                className=""
              >
                <a
                  href="/testAt"
                  onClick={[Function]}
                >
                  <strong>
                    test-name
                  </strong>
                </a>
                <span
                  className="text-gray-500"
                >
                  @
                  testAt
                  ·
                  testDat
                </span>
              </p>
              <p>
                testBody
              </p>
              <img
                alt=""
                className="w-80 pt-2 max-h-72 rounded-xl"
                data-testid="media"
                src="test-media"
              />
            </div>
          </div>
          <div
            className="ml-10"
          >
            <div
              className="flex"
            >
              <div
                className="comment-icon"
              >
                <svg
                  data-testid="commentButton"
                  fill="currentColor"
                  height="1em"
                  onClick={[Function]}
                  stroke="currentColor"
                  strokeWidth="0"
                  style={
                    Object {
                      "color": undefined,
                    }
                  }
                  viewBox="0 0 512 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  <p
    className="pl-4 pt-4"
  >
    Loading...
  </p>,
  <p>
    Error!
  </p>,
  <div />,
]
`;
